// @graphql-inspector/core/dist/ast/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
function readDocument(source) {
    var result = {
        source: source,
        fragments: [],
        operations: [],
        hasFragments: false,
        hasOperations: false,
    };
    var documentNode = graphql_1.parse(source.body);
    var filepath = source.name;
    var definitions = documentNode.definitions || [];
    definitions.forEach(function (node) {
        if (isOperation(node)) {
            result.operations.push({
                node: node,
                source: filepath,
            });
        }
        else if (isFragment(node)) {
            result.fragments.push({
                node: node,
                source: filepath,
            });
        }
    });
    result.hasFragments = result.fragments.length > 0;
    result.hasOperations = result.operations.length > 0;
    return result;
}
exports.readDocument = readDocument;
function isOperation(node) {
    return node.kind === graphql_1.Kind.OPERATION_DEFINITION;
}
function isFragment(node) {
    return node.kind === graphql_1.Kind.FRAGMENT_DEFINITION;
}
